Warn when an unknown config key is set

AliabbasMerchant opened the pull request:
## Summary

closes #855

AliabbasMerchant committed:
Warn when an unknown config key is set

mislav commented on pkg/cmd/config/config.go:
```diff
@@ -8,19 +8,31 @@ import (
 	"github.com/spf13/cobra"
 )
 
+type ConfigKey struct {
```
Minor: I would say that this type need not be exported

AliabbasMerchant replied:
Cool, I will change it (I realised it while coding, but thought we were not exactly following the exported-unexported convention, so did not unexport it :sweat_smile: )

mislav reviewed (approved):
This is great! Thank you 🙇

samcoe commented on pkg/cmd/config/config.go:
```diff
@@ -85,6 +97,18 @@ func NewCmdConfigSet(f *cmdutil.Factory) *cobra.Command {
 			}
 
 			key, value := args[0], args[1]
+			knownKey := false
+			for _, configKey := range respectedConfigKeys {
+				if key == configKey.name {
+					knownKey = true
+					break
+				}
+			}
+			if !knownKey {
+				iostreams := f.IOStreams
+				warningIcon := iostreams.ColorScheme().WarningIcon()
+				fmt.Fprintf(iostreams.ErrOut, "%s warning: '%s' is not a known configuration key\n", warningIcon, key)
+			}
```
What are your thoughts on moving this code into `cfg.Set`? We have another PR that is doing config value validation (https://github.com/cli/cli/pull/2138) and I think it would be nice to have both the key and value validations residing in the same place.

mislav replied:
I'm actually fine with this being here and not within `Set()` since it's only a warning. Other potential callers of the `Set()` function should be able to set arbitrary keys without it resulting in a warning printed on stderr.

Although, I do agree with @samcoe that this check is similar to config value validation and could thus live in the `config` package. Perhaps for now, the logic that searches for a key being among `respectedConfigKeys` could be moved to the `config` package and invoked from here like `config.IsSupportedConfig(key)`.

AliabbasMerchant replied:
Okay. So I think lets wait for #2138 to get merged, then we can decide how we need to go forward with this. I agree that it would be better to keep the validations together, now that we are validating the values also

samcoe replied:
Just wanted to update here that I merged in #2138 👍

mislav commented:
#2138 is merged! You are welcome to bring `trunk` into this branch if you want to make adjustments for it, but personally I'm not really concerned about the current duplication of where we perform these checks, so if this was merged as-is, we could always refactor down the line when we feel the need.

AliabbasMerchant commented:
Cool, thanks for the update, @samcoe and @mislav !

@mislav I am fine with anything as such, but I do believe it would be easier if we keep the checks in one common place.  

(I guess we can go forward with what @samcoe feels best amongst these two options? Majority wins? :smile: )

AliabbasMerchant commented:
So what should I do, @mislav @samcoe ?
Should I edit this, or is this going to be directly merged?

mislav committed:
Merge remote-tracking branch 'origin' into fix855

mislav commented:
@AliabbasMerchant Leave as-is; please! Thank you for your contribution so far 🙏

mislav committed:
Centralize all known configuration options

mislav committed:
Improve `git config set git_protocol` example

It's better to set this per-host in general, since setting the top-level
`git_protocol` key will not have any effect if the per-host setting
exists, and it does by default for anyone who ran `auth login`.
